### PR TITLE
modularize the crawler and decouple code  Jonathan Chun committed

### PR DESCRIPTION
**Pull request by jonchun**
created a fetcher class to move http logic away from the core crawler.

Following up on #7

### CODE REVIEW

File: badlinkfinder/fetcher.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+            min_attributes = ('scheme', 'netloc')
+            if not all([getattr(parsed_link, attr) for attr in min_attributes]):
+                logger.info('invalid link: {}'.format(current_link))
+                continue
+            neighbors.append(current_link)
+        return neighbors
+
+# Logic to figure out how to crawl the page. We don't want to search for links on assets or external sites
+def smart_request_type(url, domain=None):
```
Comment (sbrinkerhoff) [inline badlinkfinder/fetcher.py]: This logic should live in the `crawler`... the Fetcher should do what its told.

File: badlinkfinder/fetcher.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+            logger.info('URL not HTTP: {}'.format(url))
+            return False
+
+        request_type = 'GET'
+
+        """
+        Don't go crawling the entire internet. Let's stay within the original domain (if provided). However, we still want to crawl w/ HEAD request
+        just to see if the resource is even available or if there's some type of status code problem. 
+        """
+        if domain and parsed_url.netloc != domain:
```
Comment (sbrinkerhoff) [inline badlinkfinder/fetcher.py]: This should live in the crawler.

File: badlinkfinder/fetcher.py (hunk 1 above)
Comment (jonchun) [reply badlinkfinder/fetcher.py]: agreed

File: badlinkfinder/fetcher.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+            if mime_guess:
+                for asset_type in asset_types:
+                    if asset_type in mime_guess:
+                        request_type = 'HEAD'
+                        break
+
+        logger.debug('smart_request_type: {} ({})'.format(url, request_type))
+        return request_type
+
+class SiteNode:
```
Comment (sbrinkerhoff) [inline badlinkfinder/fetcher.py]: Isn't a "node" a component of a graph?  The fetcher should have no knowledge that its data is being stored in a graph, xml file, database, etc.

File: badlinkfinder/crawler.py [hunk 4]
```diff
@@ -1,16 +1,19 @@
...
 from badlinkfinder.taskqueue import TaskQueue
+import badlinkfinder.url_finder as url_finder
 
-from urllib.parse import urlparse, urlunparse
-import mimetypes
-import requests
+logger = logging.getLogger('blf.crawler')
 
 class Crawler:    
     def __init__(self, args):
```
Comment (sbrinkerhoff) [inline badlinkfinder/crawler.py]: I'd be more explicit about what Crawler needs, so it can be tested.

File: badlinkfinder/crawler.py [hunk 5]
```diff
@@ -21,139 +24,57 @@ def __init__(self, args):
...
-        if args.ignore_prefixes:
-            self.ignore_prefixes.extend(args.ignore_prefixes)
-            self.ignore_prefixes = list(dict.fromkeys(self.ignore_prefixes))
+        self.ignore_schemes = ['mailto', 'tel', 'data', 'about']
+        if args.ignore_schemes:
+            self.ignore_schemes.extend(args.ignore_schemes)
+            self.ignore_schemes = list(dict.fromkeys(self.ignore_schemes))
 
-        # verbosity must be set before logger since logger references it.
         self.verbosity = args.verbosity
```
Comment (sbrinkerhoff) [inline badlinkfinder/crawler.py]: This should be part of the CLI, where it sets the display verbosity.

**Pull request merged**